{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2f09c113-b356-4df8-8057-53bd767624db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: cryptography in c:\\users\\user\\desktop\\imageclassification\\imageclassification\\lib\\site-packages (43.0.0)\n",
      "Requirement already satisfied: cffi>=1.12 in c:\\users\\user\\desktop\\imageclassification\\imageclassification\\lib\\site-packages (from cryptography) (1.16.0)\n",
      "Requirement already satisfied: pycparser in c:\\users\\user\\desktop\\imageclassification\\imageclassification\\lib\\site-packages (from cffi>=1.12->cryptography) (2.22)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "[notice] A new release of pip is available: 23.2.1 -> 24.1.2\n",
      "[notice] To update, run: C:\\Users\\User\\Desktop\\ImageClassification\\imageclassification\\Scripts\\python.exe -m pip install --upgrade pip\n"
     ]
    }
   ],
   "source": [
    "pip install cryptography"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4f967beb-d964-46d3-bbfe-a18725be36bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The re-identified image is stored at: C:/Users/User/Desktop/dataset_creation/reidentified_valid_images\\reidentified_06_06_2024_00_05_28_6.6.2024.0.5.38_3.tif\n",
      "The re-identified image is stored at: C:/Users/User/Desktop/dataset_creation/reidentified_valid_images\\reidentified_06_06_2024_00_31_17_6.6.2024.0.31.27_1.tif\n",
      "The re-identified image is stored at: C:/Users/User/Desktop/dataset_creation/reidentified_valid_images\\reidentified_06_06_2024_00_05_28_6.6.2024.0.5.38_2.tif\n",
      "The re-identified image is stored at: C:/Users/User/Desktop/dataset_creation/reidentified_valid_images\\reidentified_06_06_2024_00_05_28_6.6.2024.0.5.38_1.tif\n",
      "The re-identified image is stored at: C:/Users/User/Desktop/dataset_creation/reidentified_valid_images\\reidentified_06_06_2024_01_42_28_6.6.2024.1.42.34_2.tif\n",
      "The re-identified image is stored at: C:/Users/User/Desktop/dataset_creation/reidentified_valid_images\\reidentified_06_06_2024_00_31_17_6.6.2024.0.31.27_2.tif\n",
      "The re-identified image is stored at: C:/Users/User/Desktop/dataset_creation/reidentified_valid_images\\reidentified_06_06_2024_01_42_28_6.6.2024.1.42.34_1.tif\n",
      "The re-identified image is stored at: C:/Users/User/Desktop/dataset_creation/reidentified_valid_images\\reidentified_06_06_2024_01_42_28_6.6.2024.1.42.34_3.tif\n",
      "The re-identified image is stored at: C:/Users/User/Desktop/dataset_creation/reidentified_valid_images\\reidentified_06_06_2024_01_42_28_6.6.2024.1.42.34_4.tif\n",
      "The re-identified image is stored at: C:/Users/User/Desktop/dataset_creation/reidentified_valid_images\\reidentified_30_07_2024_00_41_05_7.30.2024.0.41.12_1.tif\n",
      "The re-identified image is stored at: C:/Users/User/Desktop/dataset_creation/reidentified_valid_images\\reidentified_06_06_2024_01_42_28_6.6.2024.1.42.34_5.tif\n",
      "The re-identified image is stored at: C:/Users/User/Desktop/dataset_creation/reidentified_valid_images\\reidentified_30_07_2024_00_41_05_7.30.2024.0.41.12_2.tif\n",
      "The re-identified image is stored at: C:/Users/User/Desktop/dataset_creation/reidentified_valid_images\\reidentified_30_07_2024_00_41_05_7.30.2024.0.41.12_4.tif\n",
      "The re-identified image is stored at: C:/Users/User/Desktop/dataset_creation/reidentified_valid_images\\reidentified_30_07_2024_00_41_05_7.30.2024.0.41.12_3.tif\n",
      "The re-identified image is stored at: C:/Users/User/Desktop/dataset_creation/reidentified_valid_images\\reidentified_30_07_2024_00_41_05_7.30.2024.0.41.12_5.tif\n",
      "The re-identified image is stored at: C:/Users/User/Desktop/dataset_creation/reidentified_valid_images\\reidentified_30_07_2024_00_41_05_7.30.2024.0.41.12_6.tif\n",
      "The re-identified image is stored at: C:/Users/User/Desktop/dataset_creation/reidentified_valid_images\\reidentified_30_07_2024_00_41_05_7.30.2024.0.41.12_7.tif\n"
     ]
    }
   ],
   "source": [
    "import cv2\n",
    "import pytesseract\n",
    "import os\n",
    "import json\n",
    "from PIL import Image, ImageDraw, ImageFont\n",
    "from concurrent.futures import ThreadPoolExecutor\n",
    "import logging\n",
    "from cryptography.fernet import Fernet\n",
    "\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "pytesseract.pytesseract.tesseract_cmd = r'C:\\Program Files\\Tesseract-OCR\\tesseract.exe'\n",
    "\n",
    "key = Fernet.generate_key()\n",
    "handle_encryption = Fernet(key)\n",
    "\n",
    "all_valid_images = 'C:/Users/User/Desktop/dataset_creation/all_valid_images'\n",
    "deidentified_valid_images = 'C:/Users/User/Desktop/dataset_creation/deidentified_valid_images'\n",
    "reidentified_valid_images = 'C:/Users/User/Desktop/dataset_creation/reidentified_valid_images'\n",
    "\n",
    "def handle_image_processing(all_valid_images, deidentified_valid_images, reidentified_valid_images):\n",
    "    if not os.path.exists(deidentified_valid_images):\n",
    "        os.makedirs(deidentified_valid_images)\n",
    "    if not os.path.exists(reidentified_valid_images):\n",
    "        os.makedirs(reidentified_valid_images)\n",
    "    \n",
    "    with ThreadPoolExecutor(max_workers=4) as executor:\n",
    "        for filename in os.listdir(all_valid_images):\n",
    "            if filename.lower().endswith(('.png', '.jpg', '.jpeg', '.tif', '.bmp')):\n",
    "                executor.submit(image_processing_procedure, all_valid_images, deidentified_valid_images, reidentified_valid_images, filename)\n",
    "\n",
    "def image_processing_procedure(all_valid_images, deidentified_valid_images, reidentified_valid_images, filename):\n",
    "    valid_image_location = os.path.join(all_valid_images, filename)\n",
    "    deidentification_location = os.path.join(deidentified_valid_images, filename)\n",
    "    image_metadata_filename = os.path.splitext(filename)[0] + '.json'\n",
    "    image_metadata_location = os.path.join(reidentified_valid_images, image_metadata_filename)\n",
    "    \n",
    "    image_deidentification(valid_image_location, deidentification_location, image_metadata_location)\n",
    "    image_reidentification(deidentification_location, image_metadata_location, os.path.join(reidentified_valid_images, 'reidentified_' + filename))\n",
    "\n",
    "def image_deidentification(valid_image_location, deidentification_location, image_metadata_location):\n",
    "    try:\n",
    "        img = cv2.imread(valid_image_location)\n",
    "        if img is None:\n",
    "            logging.error(f\"There was an error while loading the img: {valid_image_location}\")\n",
    "            return\n",
    "        \n",
    "        txt_information = pytesseract.image_to_data(img, output_type=pytesseract.Output.DICT)\n",
    "        array_of_metadata = []\n",
    "        for i in range(len(txt_information['level'])):\n",
    "            x, y, width, height, text = txt_information['left'][i], txt_information['top'][i], txt_information['width'][i], txt_information['height'][i], txt_information['text'][i]\n",
    "            if text.strip():\n",
    "                array_of_metadata.append({'text': text, 'coords': (x, y, width, height)})\n",
    "                blurry_information(img, x, y, width, height)\n",
    "        \n",
    "        cv2.imwrite(deidentification_location, img)\n",
    "        store_encrypted_information(array_of_metadata, image_metadata_location)\n",
    "    except Exception as e:\n",
    "        logging.error(f\"There was an error while processing the img: {e}\")\n",
    "\n",
    "def blurry_information(img, x, y, width, height):\n",
    "    img[y:y+height, x:x+width] = cv2.GaussianBlur(img[y:y+height, x:x+width], (99, 99), 30)\n",
    "\n",
    "def store_encrypted_information(array_of_metadata, image_metadata_location):\n",
    "    encrypted_data = handle_encryption.encrypt(json.dumps(array_of_metadata).encode())\n",
    "    with open(image_metadata_location, 'wb') as f:\n",
    "        f.write(encrypted_data)\n",
    "    \n",
    "def image_reidentification(valid_image_location, image_metadata_location, deidentification_location):\n",
    "    try:\n",
    "        img = Image.open(valid_image_location)\n",
    "        draw = ImageDraw.Draw(img)\n",
    "        txt_font = ImageFont.load_default()\n",
    "\n",
    "        array_of_metadata = load_encrypted_information(image_metadata_location)\n",
    "        \n",
    "        for metadata_item in array_of_metadata:\n",
    "            text, (x, y, width, height) = metadata_item['text'], metadata_item['coords']\n",
    "            draw.rectangle([x, y, x + width, y + height], outline=\"white\", fill=\"white\")\n",
    "            draw.text((x, y), text, fill=\"black\", font=txt_font)\n",
    "        \n",
    "        img.save(deidentification_location)\n",
    "        print(f\"The re-identified image is stored at: {deidentification_location}\")\n",
    "    except Exception as e:\n",
    "        logging.error(f\"There was an error while re-identifying the img: {e}\")\n",
    "\n",
    "def load_encrypted_information(image_metadata_location):\n",
    "    with open(image_metadata_location, 'rb') as f:\n",
    "        encrypted_data = f.read()\n",
    "    decrypted_data = handle_encryption.decrypt(encrypted_data).decode()\n",
    "    return json.loads(decrypted_data)\n",
    "\n",
    "handle_image_processing(all_valid_images, deidentified_valid_images, reidentified_valid_images)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "61accd74-1ad5-4ab6-91f5-0b0fcbce3279",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:There was an error while processing the img: name 'get_txt_from_specific_region' is not defined\n",
      "ERROR:root:There was an error while re-identifying the img: \n",
      "ERROR:root:There was an error while processing the img: name 'get_txt_from_specific_region' is not defined\n",
      "ERROR:root:There was an error while re-identifying the img: \n"
     ]
    }
   ],
   "source": [
    "logging.basicConfig(level=logging.INFO)\n",
    "pytesseract.pytesseract.tesseract_cmd = r'C:\\Program Files\\Tesseract-OCR\\tesseract.exe'\n",
    "\n",
    "key = Fernet.generate_key()\n",
    "handle_encryption = Fernet(key)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    not_correctly_deidentified = 'C:/Users/User/Desktop/dataset_creation/not_correctly_deidentified'\n",
    "    second_deidentification = 'C:/Users/User/Desktop/dataset_creation/second_deidentification'\n",
    "    second_reidentification = 'C:/Users/User/Desktop/dataset_creation/second_reidentification'\n",
    "    \n",
    "    specific_zone_coords = [\n",
    "       (72, 85, 33, 11)\n",
    "    ]\n",
    "    \n",
    "def handle_image_processing2(not_correctly_deidentified, second_deidentification, second_reidentification, specific_zone_coords):\n",
    "    if not os.path.exists(second_deidentification):\n",
    "        os.makedirs(second_deidentification)\n",
    "    if not os.path.exists(second_reidentification):\n",
    "        os.makedirs(second_reidentification)\n",
    "    \n",
    "    for filename in os.listdir(not_correctly_deidentified):\n",
    "        if filename.lower().endswith(('.png', '.jpg', '.jpeg', '.tif', '.bmp')):\n",
    "            valid_image_location = os.path.join(not_correctly_deidentified, filename)\n",
    "            deidentification_location = os.path.join(second_deidentification, filename)\n",
    "            image_metadata_filename = os.path.splitext(filename)[0] + '.json'\n",
    "            image_metadata_location = os.path.join(second_reidentification, image_metadata_filename)\n",
    "\n",
    "            image_deidentification(valid_image_location, deidentification_location, image_metadata_location, specific_zone_coords)\n",
    "            image_reidentification(deidentification_location, image_metadata_location, os.path.join(second_reidentification, 'reidentified_' + filename))\n",
    "\n",
    "def image_deidentification(valid_image_location, deidentification_location, image_metadata_location, specific_zone_coords):\n",
    "    try:\n",
    "        img = cv2.imread(valid_image_location)\n",
    "\n",
    "        if img is None:\n",
    "            raise FileNotFoundError(f\"There was an error while loading the img: {valid_image_location}\")\n",
    "\n",
    "        array_of_metadata = []\n",
    "        for coord in specific_zone_coords:\n",
    "            left, top, width, height = coord\n",
    "            \n",
    "            original_text = get_txt_from_specific_region(img, left, top, width, height)\n",
    "            array_of_metadata.append({'text': original_text, 'coords': coord})\n",
    "            blurry_information(img, left, top, width, height)\n",
    "\n",
    "            cv2.rectangle(img, (left, top), (left + width, top + height), (0, 0, 0), -1)\n",
    "\n",
    "        cv2.imwrite(deidentification_location, img)\n",
    "        store_encrypted_information(array_of_metadata, image_metadata_location)\n",
    "\n",
    "        with open(image_metadata_location, 'w') as f:\n",
    "            json.dump(array_of_metadata, f)\n",
    "\n",
    "    except Exception as e:\n",
    "        logging.error(f\"There was an error while processing the img: {e}\")\n",
    "\n",
    "def blurry_information(img, left, top, width, height):\n",
    "    img = cv2.rectangle(img, (left, top), (left + width, top + height), (0, 0, 0), -1)\n",
    "    \n",
    "def store_encrypted_information(array_of_metadata, image_metadata_location):\n",
    "    encrypted_data = handle_encryption.encrypt(json.dumps(array_of_metadata).encode())\n",
    "    with open(image_metadata_location, 'wb') as f:\n",
    "        f.write(encrypted_data)\n",
    "    logging.info(f\"Saved encrypted metadata to {image_metadata_location}\")\n",
    "\n",
    "def image_reidentification(valid_image_location, image_metadata_location, deidentification_location):\n",
    "    try:\n",
    "        img = Image.open(valid_image_location)\n",
    "        draw = ImageDraw.Draw(img)\n",
    "        txt_font = ImageFont.load_default()\n",
    "        \n",
    "        \n",
    "        array_of_metadata = load_encrypted_information(image_metadata_location)\n",
    "                  \n",
    "        for metadata_item in array_of_metadata:\n",
    "            text, (left, top, width, height) = metadata_item['text'], metadata_item['coords']\n",
    "            draw.rectangle([left, top, left + width, top + height], outline=\"white\", fill=\"white\")\n",
    "            draw.text((left, top), text, fill=\"black\", font=txt_font)\n",
    "\n",
    "        img.save(image_metadata_location)\n",
    "        logging.info(f\"Re-identified image saved at: {image_metadata_location}\")\n",
    "        print(f\"The re-identified image is stored at: {image_metadata_location}\")\n",
    "\n",
    "    except Exception as e:\n",
    "        logging.error(f\"There was an error while re-identifying the img: {e}\")\n",
    "\n",
    "def load_encrypted_information(image_metadata_location):\n",
    "    with open(image_metadata_location, 'rb') as f:\n",
    "        encrypted_data = f.read()\n",
    "    decrypted_data = handle_encryption.decrypt(encrypted_data).decode()\n",
    "    return json.loads(decrypted_data)\n",
    "    \n",
    "handle_image_processing2(not_correctly_deidentified, second_deidentification, second_reidentification, specific_zone_coords)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "imageclassification",
   "language": "python",
   "name": "imageclassification"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
